pkg/storage/wal/manager.go

```diff
@@ -0,0 +1,123 @@
+package wal
+
+import (
+	"container/list"
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/grafana/loki/v3/pkg/logproto"
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+var (
+	ErrFull = errors.New("The WAL is full")
+)
+
+type AppendAnswer struct {
+	done chan struct{}
+	err  error
+}
+
+// Done returns a channel that is closed when the append is flushed.
+// Err() returns a non-nil error if the flush failed.
+func (p *AppendAnswer) Done() <-chan struct{} {
+	return p.done
+}
+
+// Err returns a non-nil error if the append failed. It should not
+// be called until Done() is closed to avoid data races.
+func (p *AppendAnswer) Err() error {
+	return p.err
+}
+
+type item struct {
+	r *AppendAnswer
+	w *SegmentWriter
+}
+
+type ClosedItem struct {
+	Result *AppendAnswer
+	Writer *SegmentWriter
+}
+
+type Manager struct {
+	// available is a list of items that are accepting writes. When an item
+	// is full, or the close timer has expired, it is moved to the closed list
+	// to be flushed.
+	available *list.List
+
+	// closed is a list of items that are full and waiting to be flushed.
+	// Once flushed, the item is moved back to the available list to accept
+	// writes again.
+	closed *list.List
+	mu     sync.Mutex
+}
+
+func NewManager(n int) (*Manager, error) {
+	m := Manager{
+		closed:    list.New(),
+		available: list.New(),
+	}
+	for i := 0; i < n; i++ {
+		w, err := NewWalSegmentWriter()
+		if err != nil {
+			return nil, err
+		}
+		m.available.PushBack(item{
+			r: &AppendAnswer{done: make(chan struct{})},
+			w: w,
+		})
+	}
+	return &m, nil
+}
+
+func (m *Manager) Append(_ context.Context, tenantID, labelstr string, labels labels.Labels, entries []*logproto.Entry) (*AppendAnswer, error) {
+	m.mu.Lock()
```

Comment:
Really like this idea but I'm mildly concerned about this lock - it effectively makes all Appends single threaded (unless I'm misunderstanding it!) which would potentially limit the throughput. 
Would love to see some benchmarks to see if this is worth worrying about!

Reply:
Sounds good! If it's a problem we can replace it with either a RWMutex or a Semaphore.

Reply:
Oh no we can't because we need to move the items between lists...

Reply:
As discussed offline, I don't think this will be much of a bottleneck, so as long as we can handle a reasonable QPS here, I'm more than happy to go with it! We can always identify via profiling if this needs more work.